### PR TITLE
Move from CLAClassifier to proper sequence classification 

### PR DESCRIPTION
**Pull request by marionleborgne**
@BoltzmannBrain Remove # steps (JIRA RES-53) in SequenceClassificationRegion -- adapted from CLAClassifier. Because we only want to do classification and not multistep prediction.

### CODE REVIEW

**Comment by marionleborgne** [comment]
@BoltzmannBrain This is for JIRA RES-24 `modify classification mechanism to classify only predicted active states`. Relevant changes:
* Cleaned up the classification network and made params more explicit for `createNetwork()`. 
* Factored `createRegions()` in `createNetwork()` because it was very redundant with create network.
* I need to have a nupic PR [1] merged for it to work (the TM doesn't output predictiveActiveCells when using the `tm_py` implementation)

[1] https://github.com/numenta/nupic/issues/2405

File: classification/algorithms/SequenceClassifier.py [hunk 1]
```diff
@@ -451,172 +366,32 @@ def compute(self, recordNum, patternNZ, classification, learn, infer):
...
-          key = (bit, nSteps)
-          history = self._activeBitHistory.get(key, None)
-          if history is None:
-            history = self._activeBitHistory[key] = BitHistory(self,
-                        bitNum=bit, nSteps=nSteps)
+        # Get the history structure for this bit 
+        history = self._activeBitHistory.get(bit, None)
+        if history is None:
+          history = self._activeBitHistory[bit] = BitHistory(self,
+                      bitNum=bit)
```
Comment (BoltzmannBrain) [inline classification/algorithms/SequenceClassifier.py]: indent is off

File: classification/classification_network.py [hunk 2]
```diff
@@ -225,15 +209,20 @@ def createTemporalMemoryRegion(network, prevRegionWidth):
   return temporalMemoryRegion
 
 
-def createClassifierRegion(network, classifierType, classifierParams, prevRegionWidth):
+def createClassifierRegion(network, classifierType, classifierParams, classifyPredictedActiveCells):
```
Comment (BoltzmannBrain) [inline classification/classification_network.py]: indent the continued line

File: classification/classification_network.py [hunk 3]
```diff
@@ -225,15 +209,20 @@ def createTemporalMemoryRegion(network, prevRegionWidth):
...
   """
   Create classifier region.
 
-  @param network          (Network)   The region will be a node in this network.
-  @param classifierType   (str)           Specific type of region, e.g.
-      "py.CLAClassifierRegion"; possible options can be found in /nupic/regions/.
-  @param prevRegionWidth  (int)       Width of region below.
-  @return                 (Region)    Classifier region of the network.
+  @param network          (Network)
```
Comment (BoltzmannBrain) [inline classification/classification_network.py]: The previous docstring style is much cleaner...

File: classification/classification_network.py [hunk 4]
```diff
@@ -241,108 +230,132 @@ def createClassifierRegion(network, classifierType, classifierParams, prevRegion
...
+    "classifier", classifierType, json.dumps(classifierParams))
 
   # Disable learning for now (will be enabled in a later training phase)... why???
   classifierRegion.setParameter("learningMode", False)
-  
+
   # Okay to always leave inference mode on; only there for some corner cases.
   classifierRegion.setParameter("inferenceMode", True)
 
+  if classifyPredictedActiveCells:
```
Comment (BoltzmannBrain) [inline classification/classification_network.py]: if `classifyPredictedActiveCells` is a bool you don't need this if-else statement, just do `classifierRegion.setParameter("classifyPredictedActiveCells", classifyPredictedActiveCells)`

File: classification/classification_network.py [hunk 5]
```diff
@@ -241,108 +230,132 @@ def createClassifierRegion(network, classifierType, classifierParams, prevRegion
...
-      network, sensorType, encoders, dataSource, numCats)
+  sensorOutputWidth = sensor.encoder.getWidth()
+  spInputWidth = sp.getSelf().inputWidth
+  spOutputWidth = sp.getSelf().columnCount
+  tmInputWidth = tm.getSelf().columnCount
+  tmOutputWidth = tmInputWidth * tm.getSelf().cellsPerColumn
 
-  sp = createSpatialPoolerRegion(network, sensor.encoder.getWidth())
+  if sensorOutputWidth != spInputWidth:
+    raise ValueError("Region widths do not fit. Sensor output width = %s. SP input width = %s"
```
Comment (BoltzmannBrain) [inline classification/classification_network.py]: indents are off in these ValueError statements, and better to use `.format()` for string values

File: classification/classification_network.py [hunk 6]
```diff
@@ -241,108 +230,132 @@ def createClassifierRegion(network, classifierType, classifierParams, prevRegion
...
-    sensorType   (str)          Specific type of region, e.g. "py.RecordSensor";
-                                possible options can be found in nupic/regions/.
-    encoders     (dict)         See createEncoder() docstring for format.
-    numCategories  (int)        Max number of categories of the input data.
-    classifierType   (str)      Specific type of classifier region, e.g. "py.SequenceClassifier";
-                                possible options can be found in nupic/regions/.
-    classifierParams   (dict)   Parameters for the model. E.g. {'maxCategoryCount': 3}                               
-                                
-  @return        (Network)      sensor -> SP -> TM -> CLA classifier
+  @param dataSource   (RecordStream)
```
Comment (BoltzmannBrain) [inline classification/classification_network.py]: As in `createClassifierRegion()`, the spacing in this docstring makes it difficult to read.

File: classification/classify_sensor_data.py [hunk 7]
```diff
@@ -156,30 +153,41 @@ def _setupScalarEncoder(minval, maxval):
...
-    dataSource = FileRecordStream(streamID=inputFile)
-    encoders = {"white_noise": _SCALAR_ENCODER_PARAMS}
-    network = createNetwork((dataSource, "py.RecordSensor", encoders, NUM_CATEGORIES, "py.SequenceClassifierRegion", _CLA_CLASSIFIER_PARAMS))
-
-    # Need to init the network before it can run.
-    network.initialize()
-    run(network, NUM_RECORDS, PARTITIONS, _OUT_FILE)
+  for classifyPredictedActiveCells in [True, False]:
+    for noiseAmplitude in WHITE_NOISE_AMPLITUDE_RANGES:
+      expParams = "RUNNING EXPERIMENT WITH PARAMS: \n" \
```
Comment (BoltzmannBrain) [inline classification/classify_sensor_data.py]: Use implicit continuation with parenthesis, not backslashes.

File: classification/classify_sensor_data.py [hunk 8]
```diff
@@ -156,30 +153,41 @@ def _setupScalarEncoder(minval, maxval):
...
+                  " * classifyPredictiveActiveCells=%s \n" %classifyPredictedActiveCells + \
+                  " * numRecords=%s \n" %NUM_RECORDS + \
+                  " * noiseAmplitude=%s \n" %noiseAmplitude + \
+                  " * signalAmplitude=%s \n" %SIGNAL_AMPLITUDE + \
+                  " * signalMean=%s \n" %SIGNAL_MEAN + \
+                  " * signalPeriod=%s \n" %SIGNAL_PERIOD
+      _OUT_FILE.write(expParams)
```
Comment (BoltzmannBrain) [inline classification/classify_sensor_data.py]: Always use `with open():` as a context manager for files. If something fails before you call `_OUT_FILE.close()`, the file never gets closed!

File: classification/regions/SequenceClassifierRegion.py [hunk 9]
```diff
@@ -94,7 +104,18 @@ def getSpec(cls):
             constraints='bool',
             defaultValue=1,
             accessMode='ReadWrite'),
-
+          
+          classifyPredictedActiveCells=dict(
+            description='Boolean (0/1) indicating whether or not to classify only'
+                        'predicted active cells. '
+                        'If set to 0, then all active cells will be classified.'
+                        'If set to 1, then only predicted active cells will be classified.',
```
Comment (BoltzmannBrain) [inline classification/regions/SequenceClassifierRegion.py]: over 80 characters

File: classification/regions/SequenceClassifierRegion.py [hunk 10]
```diff
@@ -235,14 +244,18 @@ def compute(self, inputs, outputs):
...
+    else:
+      activeCells = inputs["bottomUpIn"]
+      patternNZ = activeCells.nonzero()[0]
 
     # Call classifier
     clResults = self._classifier.compute(
         recordNum=self.recordNum, patternNZ=patternNZ, classification=classificationIn, learn=self.learningMode, infer=self.inferenceMode)
 
-    inferredValue = clResults["actualValues"][clResults[int(self.steps)].argmax()]
+    inferredValue = clResults["actualValues"][clResults['probabilities'].argmax()]
```
Comment (BoltzmannBrain) [inline classification/regions/SequenceClassifierRegion.py]: use double-quotes

File: classification/regions/SequenceClassifierRegion.py [hunk 11]
```diff
@@ -291,6 +307,14 @@ def getOutputElementCount(self, name):
...
     else:
       raise Exception("Unknown output {}.".format(name))
+    
+  def getInputElementCount(self, name):
+    """Returns the width of dataOut."""
+   
+    if name == "classificationResult":
+      return 1
+    else:
+      raise Exception("Unknown output {}.".format(name))
```
Comment (BoltzmannBrain) [inline classification/regions/SequenceClassifierRegion.py]: Can we use a more specific exception here?

**Comment by BoltzmannBrain** [comment]
@marionleborgne the PR you linked to is a closed issue... so without being able to test the code I'm done with this round of review.

**Comment by marionleborgne** [comment]
Thanks @BoltzmannBrain here is the PR https://github.com/numenta/nupic/pull/2411 linked to the correct issue (the other one got closed as a duplicate)

**Comment by marionleborgne** [comment]
@BoltzmannBrain The nupic PR has been merged https://github.com/numenta/nupic/pull/2411

File: classification/classify_sensor_data.py (hunk 8 above)
Comment (marionleborgne) [inline classification/classify_sensor_data.py]: fixed

File: classification/classification_network.py (hunk 5 above)
Comment (marionleborgne) [inline classification/classification_network.py]: @BoltzmannBrain  sorry but i don't see where the indents are off? can you paste here what you would like to see?

File: classification/algorithms/SequenceClassifier.py (hunk 1 above)
Comment (marionleborgne) [inline classification/algorithms/SequenceClassifier.py]: i don't this it is. the indent seems correct to me. the outer loop `for nStep in self.steps` has been removed on purpose here

**Comment by marionleborgne** [comment]
@BoltzmannBrain Thanks for the comments. i'm done with my changes. Let me know if you have any additional feedback.

File: classification/regions/SequenceClassifierRegion.py [hunk 12]
```diff
@@ -226,25 +222,45 @@ def compute(self, inputs, outputs):
 
     """
 
+    # Allow training on multiple categories:
+    #  An input can potentially belong to multiple categories. 
+    #  If a category value is < 0, it means that the input does not belong to that category.
```
Comment (BoltzmannBrain) [inline classification/regions/SequenceClassifierRegion.py]: wrap lines at 80 char

File: classification/classification_network.py (hunk 5 above)
Comment (BoltzmannBrain) [inline classification/classification_network.py]: raise ValueError("Region widths do not fit. Sensor output width = {}, SP "
                      "input width = {}.".format(sensorOutputWidth, spInputWidth)

File: classification/algorithms/SequenceClassifier.py (hunk 1 above)
Comment (BoltzmannBrain) [inline classification/algorithms/SequenceClassifier.py]: the continuation of line 380 is randomly spaced... change it to one line: `history = self._activeBitHistory[bit] = BitHistory(self, bitNum=bit)`

**Comment by BoltzmannBrain** [comment]
@marionleborgne done with this round of review. I left some comments, and few things:
- Your imports should all assume the same base directory. For example, SequenceClassifierRegion.py imports `from classification...` but classification_network.py imports `from regions...`. This works on your local machine b/c (I'm assuming) you have both research/ and research/classification/ on your python path. We cannot assume this will always be the case. We should assume only research/classification/ is on the path because this is the directory anyone running these files will be in (or importing from). So to be clear, the import in classification_network.py above is good, but the import in SequenceClassifierRegion.py needs to be `from algorithms.sequence_classifier_factory import SequenceClassifierFactory`. Please make sure all imports in this PR follow this logic; I think this was the only case affecting classify_sensor_data.py.
    - Also, please follow our [coding conventions for imports](https://github.com/numenta/nupic/wiki/Python-Style-Guide#import-grouping) (and in general). Otherwise it's extra work to see what you're importing.
- The copyright date on new files should be 2015.
- In the future please squash commits into logical pieces, it helps reviewing process a lot.
- What accuracies do we expect from running classify_sensor_data.py? I get 88.0. It would be good to include a validation check for your experiments.

File: classification/algorithms/SequenceClassifier.py (hunk 1 above)
Comment (marionleborgne) [inline classification/algorithms/SequenceClassifier.py]: oh i see what you mean now. fixed.

File: classification/classification_network.py (hunk 5 above)
Comment (marionleborgne) [inline classification/classification_network.py]: thanks. fixed.

File: classification/classification_network.py (hunk 3 above)
Comment (marionleborgne) [inline classification/classification_network.py]: fixed. using the previous docstring style again :)

File: classification/classification_network.py (hunk 4 above)
Comment (marionleborgne) [inline classification/classification_network.py]: i removed this notion of classifyPredictedActiveCells for now. i am going to tweak the tm states classification differently, after having a discussion with subutai

File: classification/classification_network.py (hunk 6 above)
Comment (marionleborgne) [inline classification/classification_network.py]: fixed as well. back to previous docstring style

File: classification/classify_sensor_data.py (hunk 7 above)
Comment (marionleborgne) [inline classification/classify_sensor_data.py]: fixed

File: classification/regions/SequenceClassifierRegion.py (hunk 9 above)
Comment (marionleborgne) [inline classification/regions/SequenceClassifierRegion.py]: fixed

File: classification/regions/SequenceClassifierRegion.py (hunk 10 above)
Comment (marionleborgne) [inline classification/regions/SequenceClassifierRegion.py]: fixed

**Comment by marionleborgne** [comment]
@BoltzmannBrain thanks for your comments. I have made the appropriate changes. Let me know if there is anything else.

**Comment by cogmission** [comment]
Interesting! Is this to be CLAClassifier The-Next-Generation? :-)

**Comment by marionleborgne** [comment]
@cogmission hello :) This work in nupic.research is mostly experimentations for now. But it's definitely helping to fix / improve the CLAClassifier. Now that this first step is almost out of the way, I can play around with how the TM states (or even UP states) are mapped to categories and see how this impacts classification accuracies for the sensor data "benchmark" data.

**Comment by BoltzmannBrain** [comment]
@marionleborgne left a few nitpicks. Again, please squash your commits. The last 9 could've been organized into 2 or 3.
Also, running your unit test fails for me:

        RuntimeError: Unknown parameter 'numCategories' for region 'classifier' of type 'py.CLAClassifierRegion'
        Valid parameters are:
           steps
           learningMode
           implementation
           alpha
           inferenceMode
           clVerbosity
           self


        ----------------------------------------------------------------------
        Ran 1 test in 0.038s

        FAILED (errors=1)

Are you running it against your local changes to CLAClassifierRegion?

**Comment by marionleborgne** [comment]
@BoltzmannBrain yes, good catch. I was running this against my local changes of the CLAClassifier regions (the PR has not yet been merged. See https://github.com/numenta/nupic/pull/2415)

**Comment by marionleborgne** [comment]
@BoltzmannBrain done with addressing the PR feedback - minus the squashed commits

File: classification/test_sensor_data_classification.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  """
+  Test classification results for sensor data. 
+  """
+
+  def testClassificationAccuracy(self):
+
+    """
+    Test classification accuracy for sensor data.
+    """
+
```
Comment (BoltzmannBrain) [inline classification/test_sensor_data_classification.py]: There should be a `tearDown()` method here to remove the test data files.

**Comment by BoltzmannBrain** [comment]
@marionleborgne add a `tearDown()` to the test then :+1:

**Pull request merged**